src/ir/auth_logic/universe_relation_insertion_test.cc

```diff
@@ -0,0 +1,142 @@
+      });
+  datalog::RelationDeclaration is_principal_decl(
+      "isPrincipal", false,
+      {datalog::Argument("x", datalog::ArgumentType::MakePrincipalType())});
+  datalog::RelationDeclaration is_property_decl(
+      "isProperty", false,
+      {
+          datalog::Argument("x",
+                            datalog::ArgumentType::MakeCustomType("Property")),
+      });
+  datalog::RelationDeclaration is_number_decl(
+      "isNumber", false,
+      {
+          datalog::Argument("x", datalog::ArgumentType::MakeNumberType()),
+      });
+  SaysAssertion assertion1(
+      Principal("\"DPTool\""),
+      {Assertion(ConditionalAssertion(
+          Fact({},
+               BaseFact(datalog::Predicate("hasToolOutput", {"app", "\"DP\""},
+                                           datalog::kPositive))),  // lhs
+          {BaseFact(datalog::Predicate("isApp", {"app"}, datalog::kPositive))}
+          // rhs
+          ))});
+  SaysAssertion dp_universe(
+      Principal("\"DPTool\""),
+      {Assertion(Fact({}, BaseFact(datalog::Predicate("isProperty", {"\"DP\""},
+                                                      datalog::kPositive))))});
+  SaysAssertion prin_universe(
+      Principal("\"DPTool\""),
+      {Assertion(
+          Fact({}, BaseFact(datalog::Predicate("isPrincipal", {"\"DPTool\""},
+                                               datalog::kPositive))))});
+  return Program({has_tool_output_decl, is_app_decl, is_property_decl,
+                  is_principal_decl, is_number_decl},
+                 {assertion1, dp_universe, prin_universe}, {});
+}
+
+template <typename T>
+std::vector<std::string> GetDebugVector(std::vector<T> nodes) {
+  return raksha::utils::MapIter<std::string>(
+      nodes, [](T node) { return node.DebugPrint(); });
+}
+
+struct ProgramEquivalenceTestData {
+  Program input_program;
+  Program expected_program;
+};
+
+class ProgramDebugStringEquivalenceTest
+    : public testing::TestWithParam<ProgramEquivalenceTestData> {};
+
+TEST_P(ProgramDebugStringEquivalenceTest,
+       SimpleProgramDebugStringEquivalenceTest) {
+  const auto& [input_program, expected_program] = GetParam();
+  // declarations
+  EXPECT_THAT(GetDebugVector(input_program.relation_declarations()),
+              UnorderedElementsAreArray(
+                  GetDebugVector(expected_program.relation_declarations())));
+  // says_assertions
+  EXPECT_THAT(GetDebugVector(input_program.says_assertions()),
+              UnorderedElementsAreArray(
+                  GetDebugVector(expected_program.says_assertions())));
+  // queries
+  EXPECT_THAT(
+      GetDebugVector(input_program.queries()),
+      UnorderedElementsAreArray(GetDebugVector(expected_program.queries())));
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    Something, ProgramDebugStringEquivalenceTest,
+    testing::Values(ProgramEquivalenceTestData{
+        UniverseRelationInsertion::InsertUniverseRelations(BuildTestProgram1()),
+        BuildTestProgram1Expected()}));
```

Comment:
There's probably not much need to make a test parameterized if you only instantiate it with a single argument.

Reply:
fixed by adding 2 more tests.